src/inmanta/ast/statements/assign.py

```diff
@@ -56,15 +64,43 @@ def __init__(self, items: typing.List[ExpressionStatement]) -> None:
         ReferenceStatement.__init__(self, items)
         self.items = items
 
+    def requires_emit_gradual(
+        self, resolver: Resolver, queue: QueueScheduler, resultcollector
+    ) -> typing.Dict[object, ResultVariable]:
+        # if we are in gradual mode, transform to a list of assignments instead of assignment of a list
+        # to get more accurate gradual execution
+        # temp variable is required get all heuristics right
+
```

Comment:
Since everything is statically typed and relations have a predefined arity, wouldn't it be possible to identity whether the list contains only single-value items, in which case `requires_emit_gradual` should just call `requires_emit`? I believe this could make sure we only introduce the overhead when required.

Reply:
If we would have static type inference: yes.

But we don't